### PR TITLE
Fix saving naming styles options page

### LINKED ISSUE #34210: Custom naming definition are not stored
VSF_TYPE_MARKDOWNHi, reproduction steps as attachement.

The problem description:
1. execute reproduction steps from attachement
2. Open naming options again
3. All custom naming conventions and specifications are missing

Thank you in advice

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/489466/custom-naming-definition-are-not-stored.html
VSTS ticketId: 820333_
_These are the original issue comments:_

Visual Studio Feedback System on 3/14/2019, 11:38 PM (3 days ago): <p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

Visual Studio Feedback System on 3/15/2019, 01:11 PM (2 days ago): <p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by JoeRobich**
- Allow page controls to store settings to option store before updating service
- Reload page control settings only after a cancel or save

Fixes #34210 which was introduced by https://github.com/dotnet/roslyn/pull/33523

The earlier PR didn't account for the Naming option page having very specialized load and save settings. Other option pages were updated to store their changes immediately into the OptionStore but Naming options still relied on SaveSettings being called to directly update the OptionService. In a related way, Naming service overwrites its view model when LoadSettings is called and because of the earlier change LoadSettings was being called every time the page is activated.

<details><summary>Ask Mode template</summary>

### Customer scenario

User tries to update naming styles for identifiers and changes do not save.

### Bugs this fixes

#34210

### Workarounds, if any

None

### Risk

Low

### Performance impact

Low

### Is this a regression from a previous update?

Yes, regression was introduced by https://github.com/dotnet/roslyn/pull/33523

### Root cause analysis

In Dev16, option saving was rewritten so that all option changes went to a centralized store and were saved to the option service in a single location. This rewrite missed the custom save behavior of the Naming styles option page.

### How was the bug found?

Reported through developer feedback

</details>

### CODE REVIEW

File: src/Tools/ILAsm/IlAsmDeploy.csproj [hunk 1]
```diff
@@ -12,9 +12,6 @@
     <TargetLatestRuntimePatch>false</TargetLatestRuntimePatch>
     <SelfContained>true</SelfContained>
 
-    <!-- Do not build again when publishing -->
-    <NoBuild>true</NoBuild>
```
Comment (JoeRobich) [inline src/Tools/ILAsm/IlAsmDeploy.csproj]: This is unrelated to the fix, but allows for builds in VS to not fail. This same change is already in the master branch.

**Comment by JoeRobich** [comment]
@jasonmalinowski Could you please take a look?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 2]
```diff
@@ -50,16 +52,22 @@ protected override void OnActivate(System.ComponentModel.CancelEventArgs e)
...
 
-                s_needsLoadOnNextActivate = false;
+                s_needsToUpdateOptionStore = false;
             }
 
-            pageControl.LoadSettings();
+            if (_needsLoadOnNextActivate)
+            {
+                // For pages that don't use option bindings we need to load setting changes.
+                pageControl.LoadSettings();
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Consider changing `LoadSettings` to _something_ like `LoadNonDataboundSettings`? `LoadSettingsThatDoNotUseOptionBinding`? ¯\(°_o)/¯

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 3]
```diff
@@ -81,19 +89,33 @@ public override void LoadSettingsFromStorage()
...
+            _needsLoadOnNextActivate = true;
         }
 
         public override void SaveSettingsToStorage()
         {
             EnsureOptionPageCreated();
 
+            // Allow page controls to perist their settings to the option store before updating the
+            // option service.
+            pageControl.SaveSettings();
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Same naming concern as `LoadSettings`

**Comment by dpoeschl** [comment]
Note: This has been tested by the test team. Only [one bug](https://github.com/dotnet/roslyn/issues/34215) was found, which is unrelated and has existed since 15.x.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 4]
```diff
@@ -81,19 +89,36 @@ public override void LoadSettingsFromStorage()
...
+            // Allow page controls to perist their settings to the option store before updating the
+            // option service.
+            pageControl.OnSave();
+
             // Save the changes that were accumulated in the option store.
-            s_optionService.SetOptions(s_optionStore.GetOptions());
+            var oldOptions = s_optionService.GetOptions();
+            var newOptions = s_optionStore.GetOptions();
+            s_optionService.SetOptions(newOptions);
+            OptionLogger.Log(oldOptions, newOptions);
```
Comment (JoeRobich) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: As pointed out by David offline, we had been logging option changes on SaveSettings before the earlier PR moved logging into the OptionStore. This reverts the behavior back.

**Comment by JoeRobich** [comment]
Integration test failure was already logged as https://github.com/dotnet/roslyn/issues/33652

**Pull request merged**